source/extensions/transport_sockets/tls/context_impl.h

```diff
@@ -88,6 +88,10 @@ class ContextImpl : public virtual Envoy::Ssl::Context,
   Envoy::Ssl::CertificateDetailsPtr getCaCertInformation() const override;
   std::vector<Envoy::Ssl::CertificateDetailsPtr> getCertChainInformation() const override;
   absl::optional<uint64_t> secondsUntilFirstOcspResponseExpires() const override;
+  void initSslContext(SSL_CTX* ctx) const override;
```

Comment:
I think it is leaking too many details to expose how SSL_CTX is initialized. Can you restructure this extension so that you provide the ContextImpl with a certificate, and have it create the `TlsContext` and `SSL_CTX` for it?

Reply:
The `TlsContext` is created in `ContextImpl::ContextImpl`, do you mean this extension provide certificates during `ContextImpl::ContextImpl`?
If yes, then we can not select certificate on demand, base on the incoming SNI server name.

After the extension getting a certificate(in its custom logic), it could only create a SSL_CTX with certificated/key inside.
So, it need a way to inherit some other configurations from the `Ssl::Context`, i.e. min/max version, cert validation and etc.

How about move the `initSslContext` into a new Class, i.e.
```
class ContextCallback {
  virtual initSslContext(SSL_CTX* ctx);
}
```
then, the extension only provider a ContextCallback instance to the extension's custom logic?